File: mas-mcp/documentation-manager-server/python_conversion/src/documentation_manager/services/document_service.py
import uuid
from datetime import datetime
from typing import Dict, List, Optional, Union, Any

from ..database.database_service import DatabaseService
from ..models.document import (
    Document,
    DocumentMetadata,
    DocumentGenerationRequest,
    DocumentRetrievalRequest,
    DocumentUpdateRequest,
    DocumentVersionInfo,
)


class DocumentService:
    """Service for document management operations"""

    def __init__(self, db_service: DatabaseService):
        self.db = db_service

    async def create_document(
        self, request: DocumentGenerationRequest, content: str
    ) -> Document:
        """Create a new document based on generation request"""
        now = datetime.utcnow()
        document = Document(
            id=str(uuid.uuid4()),
            type=request.type,
            title=request.title,
            description="",
            project_id=request.project_id,
            version="1.0.0",
            created_by=request.created_by,
            created_at=now,
            updated_by=request.created_by,
            updated_at=now,
            tags=request.tags or [],
            content=content,
        )

        await self.db.create_document(document)

        # Add initial version
        version_info = DocumentVersionInfo(
            id=str(uuid.uuid4()),
            document_id=document.id,
            version=document.version,
            changes="Initial creation",
            created_by=request.created_by,
            created_at=now,
            content=content,
        )

        await self.db.add_version(version_info)

        return document

    async def get_document(
        self, document_id: str, include_content: bool = True
    ) -> Optional[Document]:
        """Get a document by ID"""
        document = await self.db.get_document(document_id)
        if not document:
            return None

        if not include_content:
            # Create a copy without content
            return Document(
                id=document.id,
                type=document.type,
                title=document.title,
                description=document.description,
                project_id=document.project_id,
                version=document.version,
                created_by=document.created_by,
                created_at=document.created_at,
                updated_by=document.updated_by,
                updated_at=document.updated_at,
                tags=document.tags,
                content="",  # Empty content
            )

        return document

    async def update_document(self, request: DocumentUpdateRequest) -> Optional[Document]:
        """Update an existing document"""
        document = await self.db.get_document(request.id)
        if not document:
            return None

        now = datetime.utcnow()
        
        # Apply updates
        if request.title is not None:
            document.title = request.title
        if request.description is not None:
            document.description = request.description
        if request.tags is not None:
            document.tags = request.tags
        
        document.updated_by = request.updated_by
        document.updated_at = now

        # Update content and increment version if content changed
        content_changed = False
        if request.content is not None and request.content != document.content:
            document.content = request.content
            content_changed = True

            # Increment version
            version_parts = document.version.split(".")
            minor = int(version_parts[1]) + 1
            document.version = f"{version_parts[0]}.{minor}.0"

            # Add version
            version_info = DocumentVersionInfo(
                id=str(uuid.uuid4()),
                document_id=document.id,
                version=document.version,
                changes="Content updated",
                created_by=request.updated_by,
                created_at=now,
                content=document.content,
            )

            await self.db.add_version(version_info)

        return await self.db.update_document(document)

    async def delete_document(self, document_id: str) -> bool:
        """Delete a document"""
        return await self.db.delete_document(document_id)

    async def search_documents(
        self, query: DocumentRetrievalRequest
    ) -> List[Union[Document, DocumentMetadata]]:
        """Search for documents based on criteria"""
        return await self.db.search_documents(query)

    async def get_document_versions(self, document_id: str) -> List[DocumentVersionInfo]:
        """Get all versions of a document"""
        return await self.db.get_versions(document_id)

    async def get_document_version(
        self, document_id: str, version: str
    ) -> Optional[Document]:
        """Get a specific version of a document"""
        document = await self.db.get_document(document_id)
        if not document:
            return None

        versions = await self.db.get_versions(document_id)
        version_info = next((v for v in versions if v.version == version), None)
        if not version_info:
            return None

        # Create a document with the version data
        return Document(
            id=document.id,
            type=document.type,
            title=document.title,
            description=document.description,
            project_id=document.project_id,
            version=version_info.version,
            created_by=document.created_by,
            created_at=document.created_at,
            updated_by=version_info.created_by,
            updated_at=version_info.created_at,
            tags=document.tags,
            content=version_info.content,
        ) 

File: mas-mcp/documentation-manager-server/src/services/documentService.ts
import { 
  Document, 
  DocumentMetadata, 
  DocumentGenerationRequest, 
  DocumentRetrievalRequest,
  DocumentUpdateRequest,
  DocumentVersionInfo
} from '../types.js';

// This will be replaced with actual database implementation
class DocumentDatabaseService {
  private documents: Map<string, Document> = new Map();
  private versions: Map<string, DocumentVersionInfo[]> = new Map();

  async createDocument(document: Document): Promise<Document> {
    this.documents.set(document.id, document);
    if (!this.versions.has(document.id)) {
      this.versions.set(document.id, []);
    }
    return document;
  }

  async getDocument(id: string): Promise<Document | null> {
    return this.documents.get(id) || null;
  }

  async updateDocument(document: Document): Promise<Document> {
    this.documents.set(document.id, document);
    return document;
  }

  async deleteDocument(id: string): Promise<boolean> {
    return this.documents.delete(id);
  }

  async searchDocuments(query: DocumentRetrievalRequest): Promise<Document[]> {
    const results: Document[] = [];
    for (const doc of this.documents.values()) {
      if (this.matchesQuery(doc, query)) {
        results.push(doc);
      }
    }
    return results.slice(query.offset || 0, (query.offset || 0) + (query.limit || results.length));
  }

  async addVersion(version: DocumentVersionInfo): Promise<void> {
    const versions = this.versions.get(version.documentId) || [];
    versions.push(version);
    this.versions.set(version.documentId, versions);
  }

  async getVersions(documentId: string): Promise<DocumentVersionInfo[]> {
    return this.versions.get(documentId) || [];
  }

  private matchesQuery(doc: Document, query: DocumentRetrievalRequest): boolean {
    if (query.id && doc.id !== query.id) return false;
    if (query.projectId && doc.projectId !== query.projectId) return false;
    if (query.type && doc.type !== query.type) return false;
    if (query.version && doc.version !== query.version) return false;
    if (query.tags && query.tags.length > 0) {
      if (!query.tags.some(tag => doc.tags.includes(tag))) {
        return false;
      }
    }
    return true;
  }
}

export class DocumentService {
  private db: DocumentDatabaseService;
  
  constructor(dbService?: DocumentDatabaseService) {
    this.db = dbService || new DocumentDatabaseService();
  }

  async createDocument(request: DocumentGenerationRequest, content: string): Promise<Document> {
    const now = new Date();
    const document: Document = {
      id: crypto.randomUUID(),
      type: request.type,
      title: request.title,
      description: '',
      projectId: request.projectId,
      version: '1.0.0',
      createdBy: request.createdBy,
      createdAt: now,
      updatedBy: request.createdBy,
      updatedAt: now,
      tags: request.tags || [],
      content
    };

    await this.db.createDocument(document);
    
    // Add initial version
    const versionInfo: DocumentVersionInfo = {
      id: crypto.randomUUID(),
      documentId: document.id,
      version: document.version,
      changes: 'Initial creation',
      createdBy: request.createdBy,
      createdAt: now,
      content
    };
    
    await this.db.addVersion(versionInfo);
    
    return document;
  }

  async getDocument(id: string, includeContent = true): Promise<Document | null> {
    const document = await this.db.getDocument(id);
    if (!document) return null;
    
    if (!includeContent) {
      const { content, ...metadata } = document;
      return { ...metadata, content: '' } as Document;
    }
    
    return document;
  }

  async updateDocument(request: DocumentUpdateRequest): Promise<Document | null> {
    const document = await this.db.getDocument(request.id);
    if (!document) return null;

    const now = new Date();
    const updatedDocument: Document = {
      ...document,
      title: request.title || document.title,
      description: request.description || document.description,
      tags: request.tags || document.tags,
      updatedBy: request.updatedBy,
      updatedAt: now,
      content: request.content !== undefined ? request.content : document.content
    };

    // Increment version if content changed
    if (request.content !== undefined && request.content !== document.content) {
      const versionParts = document.version.split('.');
      const minor = parseInt(versionParts[1] || '0') + 1;
      updatedDocument.version = `${versionParts[0]}.${minor}.0`;
      
      // Add version
      const versionInfo: DocumentVersionInfo = {
        id: crypto.randomUUID(),
        documentId: document.id,
        version: updatedDocument.version,
        changes: 'Content updated',
        createdBy: request.updatedBy,
        createdAt: now,
        content: updatedDocument.content
      };
      
      await this.db.addVersion(versionInfo);
    }

    return this.db.updateDocument(updatedDocument);
  }

  async deleteDocument(id: string): Promise<boolean> {
    return this.db.deleteDocument(id);
  }

  async searchDocuments(query: DocumentRetrievalRequest): Promise<Document[]> {
    const documents = await this.db.searchDocuments(query);
    
    if (!query.fullText) {
      return documents.map(doc => {
        const { content, ...metadata } = doc;
        return { ...metadata, content: '' } as Document;
      });
    }
    
    return documents;
  }

  async getDocumentVersions(documentId: string): Promise<DocumentVersionInfo[]> {
    return this.db.getVersions(documentId);
  }

  async getDocumentVersion(documentId: string, version: string): Promise<Document | null> {
    const document = await this.db.getDocument(documentId);
    if (!document) return null;
    
    const versions = await this.db.getVersions(documentId);
    const versionInfo = versions.find(v => v.version === version);
    if (!versionInfo) return null;
    
    return {
      ...document,
      version: versionInfo.version,
      content: versionInfo.content,
      updatedAt: versionInfo.createdAt,
      updatedBy: versionInfo.createdBy
    };
  }
} 